src/main/java/io/cdap/plugin/http/source/common/pagination/page/BasePage.java

```diff
@@ -42,14 +56,78 @@ public BasePage(BaseHttpSourceConfig config, HttpErrorHandler httpErrorHandler,
 
   @Nullable
   public abstract String getPrimitiveByPath(String path);
+  public abstract boolean hasNext();
+  protected abstract StructuredRecord next() throws TransformationException;
 
   public HttpErrorHandlingStrategy getErrorHandlingStrategy() {
     return httpErrorHandler
       .getErrorHandlingStrategy(httpResponse.getStatusCode())
       .getAfterRetryStrategy();
   }
 
-  public HttpResponse getHttpResponse() {
-    return httpResponse;
+  /**
+   * Run functions for every record on the page if appropriate. Or for whole page if has erroneous http
+   * status code.
+   *
+   * @param onOutput is run for non-erroneous records on the page
+   * @param onSendToError is run for erroneous records with "send to error" strategy. Or once for whole page with
+   *                      erroneous http status code and strategy "send to error".
+   */
+  public void consume(Consumer<StructuredRecord> onOutput, Consumer<InvalidEntry<StructuredRecord>> onSendToError)
+    throws TransformationException {
```

Comment:
this should no longer throw TransformationException

Reply:
It case the error handling strategy is "stop on error" it will throw TransformationException with detailed information due to which transformation it failed.

That's why we need this throw.